### PR TITLE
Reauth conn

### PR DESCRIPTION
**Pull request by GilboaAWS**
Make sure that:

- [ ] You have read the [contribution guidelines](https://github.com/lettuce-io/lettuce-core/blob/main/.github/CONTRIBUTING.md).
- [ ] You have created a feature request first to discuss your contribution intent. Please reference the feature request ticket number in the pull request.
- [ ] You use the code formatters provided [here](https://github.com/lettuce-io/lettuce-core/blob/main/formatting.xml) and have them applied to your changes. Don’t submit any formatting related changes.
- [ ] You submit test cases (unit or integration tests) that back your changes.

### CODE REVIEW

File: src/main/java/io/lettuce/core/AbstractRedisClient.java [hunk 1]
```diff
@@ -261,6 +267,28 @@ public void removeListener(CommandListener listener) {
...
+     *     
+     * @since 6.2
+     */
+    abstract public void reauthConnections();
+
+    /**
+     * Apply a {@link Consumer} of {@link Closeable} to all active connections.
+     *
+     * @param <T>
+     * @param function the {@link Consumer}.
```
Comment (shachlanAmazon) [inline src/main/java/io/lettuce/core/AbstractRedisClient.java:281]: add doc for selector, too

File: src/main/java/io/lettuce/core/ClientOptions.java [hunk 2]
```diff
@@ -194,6 +216,60 @@ public Builder autoReconnect(boolean autoReconnect) {
             return this;
         }
 
+        /**
+         * Enables or disables periodic reauthenticate of active connections. Defaults to {@code false}. See
```
Comment (shachlanAmazon) [inline src/main/java/io/lettuce/core/ClientOptions.java]: doc default duration.

File: src/main/java/io/lettuce/core/ClientOptions.java [hunk 3]
```diff
@@ -194,6 +216,60 @@ public Builder autoReconnect(boolean autoReconnect) {
...
+         * @return {@code this}
+         * @since 6.2
+         */
+        public Builder enablePeriodicReauthentication(boolean periodicReauthentication) {
+            this.shouldReauthenticatePeriodically = periodicReauthentication;
+            return this;
+        }
+
+        /**
+         * Enables periodic reauthenticate of active connections. Defaults to {@code false}. See
```
Comment (shachlanAmazon) [inline src/main/java/io/lettuce/core/ClientOptions.java]: same

File: src/main/java/io/lettuce/core/ClientOptions.java [hunk 4]
```diff
@@ -194,6 +216,60 @@ public Builder autoReconnect(boolean autoReconnect) {
...
+
+        /**
+         * Set the refresh period. Defaults to {@literal 60 MINUTES}. See {@link #DEFAULT_REAUTHENTICATE_PERIOD} and
+         * {@link #DEFAULT_REAUTHENTICATE_PERIOD_UNIT}.
+         *
+         * @param reauthenticationPeriod period for triggering reauthentication, must be greater {@literal 0}
+         * @return {@code this}
+         * @since 6.2
+         */
+        public Builder reauthenticationPeriod(Duration reauthenticationPeriod) {
```
Comment (shachlanAmazon) [inline src/main/java/io/lettuce/core/ClientOptions.java:268]: setReauth...?

File: src/main/java/io/lettuce/core/ClientOptions.java [hunk 5]
```diff
@@ -194,6 +216,60 @@ public Builder autoReconnect(boolean autoReconnect) {
...
+         * @param reauthenticationPeriod period for triggering reauthentication, must be greater {@literal 0}
+         * @return {@code this}
+         * @since 6.2
+         */
+        public Builder enablePeriodicReauthentication(Duration reauthenticationPeriod) {
+            return reauthenticationPeriod(reauthenticationPeriod).enablePeriodicReauthentication();
+        }
+
+        /**
+         * Set the refresh period. Defaults to {@literal 60 MINUTES}. See {@link #DEFAULT_REAUTHENTICATE_PERIOD} and
```
Comment (shachlanAmazon) [inline src/main/java/io/lettuce/core/ClientOptions.java]: refresh -> reauthentication
because you don't use the term `refresh` elsewhere for this.

File: src/main/java/io/lettuce/core/ConnectionState.java [hunk 6]
```diff
@@ -110,6 +110,10 @@ protected void setUserNamePassword(List<char[]> args) {
             return;
         }
 
+        if (!(this.credentialsProvider instanceof StaticCredentialsProvider)) {
```
Comment (shachlanAmazon) [inline src/main/java/io/lettuce/core/ConnectionState.java:117]: consider adding a doc explaining why this happens

File: src/main/java/io/lettuce/core/RedisClientReauthScheduler.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+                logger.debug("Could not cancel client reauth", e);
+            }
+        }
+    }
+
+    @Override
+    public void run() {
+        try {
+            clientToReauth.reauthConnections();
+        } catch (IllegalArgumentException e) {
```
Comment (shachlanAmazon) [inline src/main/java/io/lettuce/core/RedisClientReauthScheduler.java]: why don't you handle other exception types?

File: src/main/java/io/lettuce/core/StatefulRedisConnectionImpl.java [hunk 8]
```diff
@@ -293,4 +295,29 @@ public ConnectionState getConnectionState() {
...
+                ((RedisCredentialsProvider.ImmediateRedisCredentialsProvider)credentialsProvider).resolveCredentialsNow();
+            CharSequence passwd = CharBuffer.wrap(redisCredentials.getPassword());
+            if (redisCredentials.hasUsername()) {
+                async().auth(redisCredentials.getUsername(), passwd);
+            } else {
+                async().auth(passwd);
+            }
+        }
+
+        CompletableFuture<RedisCredentials> credentialsFuture = credentialsProvider.resolveCredentials().toFuture();
```
Comment (shachlanAmazon) [inline src/main/java/io/lettuce/core/StatefulRedisConnectionImpl.java]: should this be inside an `else` clause from the `if` in line 302?

File: src/main/java/io/lettuce/core/StatefulRedisConnectionImpl.java [hunk 9]
```diff
@@ -293,4 +295,29 @@ public ConnectionState getConnectionState() {
         return state;
     }
 
+    public void reauthenticate()
+    {
+        RedisCredentialsProvider credentialsProvider = state.getCredentialsProvider();
+
+        if (credentialsProvider instanceof RedisCredentialsProvider.ImmediateRedisCredentialsProvider) {
```
Comment (shachlanAmazon) [inline src/main/java/io/lettuce/core/StatefulRedisConnectionImpl.java:302]: why is this required? Can't the immediate provider return the correct credentialsin `resolveCredentials`?

File: src/main/java/io/lettuce/core/cluster/RedisClusterClient.java [hunk 10]
```diff
@@ -936,6 +938,30 @@ protected Partitions loadPartitions() {
...
+        });
+
+        forEachClusterPubSubConnection(input -> {
+            input.reauthenticate(creds);
+        });
+    }
+    
+    public void reauthConnections() {
+        RedisCredentialsProvider credentialsProvider = getFirstUri().getCredentialsProvider();
+        if (credentialsProvider instanceof RedisCredentialsProvider.ImmediateRedisCredentialsProvider) {
```
Comment (shachlanAmazon) [inline src/main/java/io/lettuce/core/cluster/RedisClusterClient.java:953]: same (is this required, shouldn't there be an `else` clause?

**Review by shachlanAmazon — APPROVED** [review]

File: src/main/java/io/lettuce/core/AbstractRedisClient.java (hunk 1 above)
Comment (GilboaAWS) [reply src/main/java/io/lettuce/core/AbstractRedisClient.java:281]: It's whole cropped from other class.

File: src/main/java/io/lettuce/core/StatefulRedisConnectionImpl.java (hunk 9 above)
Comment (GilboaAWS) [reply src/main/java/io/lettuce/core/StatefulRedisConnectionImpl.java:302]: Keep this behavior from other places in the code, see RedisHandshake

File: src/main/java/io/lettuce/core/cluster/RedisClusterClient.java (hunk 10 above)
Comment (GilboaAWS) [reply src/main/java/io/lettuce/core/cluster/RedisClusterClient.java:953]: the same answer as above
